Add support for customisation scripts

OddBloke opened the pull request.

Commit:
Add support for customisation scripts to _write_cloud_config

Commit:
Hook up customisation script logic in main so it's usable

Commit:
Add documentation for customisation scripts

josvazg commented:
LGTM!

Code seems fine and:

========================================================================================== 17 passed in 0.20 seconds ==========================================================================================
___________________________________________________________________________________________________ summary ___________________________________________________________________________________________________
  py27: commands succeeded
  py35: commands succeeded
  congratulations :)

josvazg reviewed (approved):
LGTM!
And tests pass!

rcj4747 commented on README.md:
````diff
@@ -100,3 +100,30 @@ To fetch, for example, the root squashfs, you can simply use `scp`:
+something like this out to `script.sh`:
+
+```
+#!/bin/sh
````
I would suggest adding "-x" to match the other output the user will see.

Also an echo won't hurt as an example of a good practice to say we're here.  The build output will have no delimiter saying that it's executing this hook so debugging will be harder for the end user without some breadcrumbs.

`echo "======================== Running customization script ==========================="`

rcj4747 replied:
After that I'm +1.  The tests pass and images boot (I tested disk1).

rcj4747 reviewed (changes requested):
Finding the execution of their custom script will be a chore the first time.  Let's have our example show use of 'set -x' and an 'echo' to say that we've arrived.

rcj4747 commented on generate_build_config.py:
```diff
@@ -33,23 +34,43 @@
 
+WRITE_FILES_TEMPLATE = """\
+write_files:
+- encoding: b64
```
I was going to suggest gz+b64 as some clouds we have seen have relatively low limits on the size of user-data.  However, I don't know of a python 2/3 compatible way to do that.  Python2 uses gzip.GzipFile on a StringIO object but Python3 would need to use io.BytesIO or maybe just str.encode('gzip').

I was going to suggest (for python2):

```
     if customisation_script is not None:
         with open(customisation_script, 'rb') as f:
-            content = base64.b64encode(f.read()).decode('utf-8')
+            content_gz = StringIO.StringIO()
+            with gzip.GzipFile(fileobj=content_gz, mode='w') as fgz:
+                fgz.write(f.read())
+            content = base64.b64encode(content_gz.getvalue()).decode('utf-8')
         if content:
             output_string += '\n'
```

rcj4747 replied:
I figured it out, thoughts?

```
diff --git a/generate_build_config.py b/generate_build_config.py
index a1e39e1..834628a 100755
--- a/generate_build_config.py
+++ b/generate_build_config.py
@@ -1,7 +1,12 @@
 #!/usr/bin/env python
 from __future__ import print_function
 
+try:
+    from StringIO import StringIO as IO
+except:
+    from io import BytesIO as IO
 import base64
+import gzip
 import sys
 
 
@@ -36,7 +41,7 @@ runcmd:
 
 WRITE_FILES_TEMPLATE = """\
 write_files:
-- encoding: b64
+- encoding: gz+b64
   content: {content}
   path:
     /home/ubuntu/build-output/chroot-autobuild/usr/share/livecd-rootfs/live-build/ubuntu-cpc/hooks/9999-local-modifications.chroot
@@ -55,7 +60,10 @@ def _write_cloud_config(output_file, customisation_script=None):
     output_string = TEMPLATE
     if customisation_script is not None:
         with open(customisation_script, 'rb') as f:
-            content = base64.b64encode(f.read()).decode('utf-8')
+            content_gz = IO()
+            with gzip.GzipFile(fileobj=content_gz, mode='w') as fgz:
+                fgz.write(f.read())
+            content = base64.b64encode(content_gz.getvalue()).decode('utf-8')
         if content:
             output_string += '\n'
             output_string += WRITE_FILES_TEMPLATE.format(content=content)
```

rcj4747 replied:
And I can't find the user-data metadata limit for Azure as that's the cloud used in the example.  Practically speaking I don't think their limit is all that low (16K maybe?).

rcj4747 replied:
And this can be another PR.  My abstention for this PR is the example customization script comments only.  The patch above was tested and produced a bootable disk1.img with customizations applied.

Commit:
Minor cleanup of incorrect README language

Commit:
Improve example customisation script (per rcj's review comments)

Specifically: https://github.com/OddBloke/ubuntu-standalone-builder/pull/6/files/6a1022d1c845fc0498a2d61992601685a11719d1#r97921388

chrisglass commented on generate_build_config.py:
```diff
@@ -33,23 +34,43 @@
     Write an image building cloud-config file to a given location.
 
     :param output_file:
         The path for the file to write to.
```
A :param entry for customisation_script is missing here.

OddBloke replied:
Fixed.

Commit:
Add documentation for customisation_script _write_cloud_config parameter
